### PR TITLE
Soen 8 communication

### PR DESCRIPTION
**Pull request by jfv-2000**
# Description

Implements/fixes [SOEN-8](https://soen390team8.atlassian.net/jira/software/projects/SOEN/boards/1?selectedIssue=SOEN-8)

- Chat functionality implemented with the use of web sockets
- Backend and db implemented to web socket to load past messages 
- Backend web sockets and routes tested, frontend testing still has to be improved
Doctor:
- May select among their list of patients to chat with

Patient:
- Message link brings the patient directly to their chat with their assigned doctor

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit testing & Integration testing for sockets

# Screenshots
image
image

### CODE REVIEW

**Review by stevenvuong00 — APPROVED** [review]
fantastic work! Love the screenshots of the dm leaks

File: frontend/components/communication/message-box.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    patient_accountId: number;
+    doctor_accountId: number;
+    firstName?: string;
+    lastName?: string;
+    doctorName?: string;
+}
+
+const MessageBox = ({ patient_accountId, doctor_accountId, firstName, lastName, doctorName }: MessageBoxProps) => {
+    const { data: session } = useSession();
+    const { messageList, send } = useWebSockets({
```
Comment (NicholasYip) [inline frontend/components/communication/message-box.tsx:17]: wow custom hooks

**Review by NicholasYip — COMMENTED** [review]
small question, rest looks good!

**Review by NicholasYip — APPROVED** [review]

File: frontend/package.json [hunk 2]
```diff
@@ -27,8 +27,10 @@
         "react-dom": "17.0.2",
         "react-icons": "^4.3.1",
         "react-responsive-carousel": "^3.2.23",
+        "react-router-dom": "^6.2.2",
         "recharts": "^2.1.9",
         "sass": "^1.49.0",
+        "socket.io-client": "^4.1.3",
```
Comment (NicholasYip) [inline frontend/package.json:33]: why is socket.io-client a required package in the backend?

Comment (jfv-2000) [reply frontend/package.json:33]: Just for the T E S T S cause to test the socket server and its functionalities, we have to use both the socket client and the socket server

Comment (NicholasYip) [reply frontend/package.json:33]: Awesome, thanks for clarifying! I love tests

Comment (jfv-2000) [reply frontend/package.json:33]: No problem ! I'm glad you're enjoying tests, I haven't done the frontend testing portion yet, so feel free to do it !

**Review by KelvinTrang — APPROVED** [review]
Chief Jeff kiss. Good job mate

**Pull request merged**